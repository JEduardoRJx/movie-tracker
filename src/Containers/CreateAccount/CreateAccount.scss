.form-section {
  display: flex;
  flex-direction: column;
}

.create-account-h2,
.error-msg {
  color: white;
}