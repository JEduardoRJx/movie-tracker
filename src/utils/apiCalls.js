const baseUrl = 'https://api.themoviedb.org/3/movie/now_playing?';
// const apiKey = '1b20ae1afe685b2871c8d94218f89eba';
const apiKey2 = 'cee1e60becdb4297de68233fbef2f560';

export const fetchMovies = async () => {
  const url = `${baseUrl}api_key=${apiKey2}&language=en-US`
  const response = await fetch(url)
    if (!response.ok) {
      throw Error("There was an issue retrieving your movies. Please try again.");
    }
  let movies = await response.json();
  movies = movies.results.map(movie => {
    const { backdrop_path, genre_ids, id, overview, poster_path, release_date, title, vote_average } = movie
    return { backdrop_path, genre_ids, id, overview, poster_path, release_date, title, vote_average, isFavorite: false }
  })
  return movies
};

export const postUser = async user => {
  const url = 'http://localhost:3001/api/v1/login';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      email: user.email,
      password: user.password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);
  if (!response.ok && response.status !== 401) {
    throw Error('Sorry, unable to retrieve your account. Try again later.')
  }
  const foundUser = await response.json();
  if (response.status === 401) {
    throw new Error('Username or password incorrect')
  }
  return foundUser;
}

export const addUser = async user => {
  const url = 'http://localhost:3001/api/v1/users';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      name: user.name,
      email: user.email,
      password: user.password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);
    if (!response.ok && response.status !== 500) {
     throw Error('Sorry, unable to create your account. Try again later.')
  }
  const newUser = await response.json();
    if (response.status === 500) {
      throw new Error('There is already an account with this email. Go to login or use another email address.')
  }
  return newUser;
}

export const addFavorite = async (userId, movie) => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  const options = {
    method: 'POST',
    body: JSON.stringify({
      movie_id: movie.id, 
      title: movie.title, 
      poster_path: movie.poster_path, 
      release_date: movie.release_date, 
      vote_average: movie.vote_average, 
      overview: movie.overview
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to add movie as favorite. Try again later.')
  }
  const addedFav = await response.json();
  return addedFav;
}

export const getFavorites = async userId => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  const response = await fetch(url)
  if(!response.ok) {
    throw Error('Unable to retrieve your favorite movies as this time. Please try again later')
  }
  const favoriteMovies = await response.json()
  return favoriteMovies
}

export const removeFavorite = async (userId, favId) => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites/${favId}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to remove this movie as a favorite at this time. Please try again later')
  }
}