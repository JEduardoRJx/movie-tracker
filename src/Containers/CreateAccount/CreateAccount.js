import React, { Component } from 'react';
import './CreateAccount.scss';
import { Link } from 'react-router-dom';
import { addUser } from '../../utils/apiCalls';


class CreateAccount extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: ''
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  createAccount = (e) => {
    addUser(this.state)
    this.clearInputs()
  }

  clearInputs = () => {
    this.setState({ name: '', email: '', password:''})
  }

  render() {
    return (
      <section className='form-section'>
        <form className='create-account-form'>
          <h2>Create Account: </h2>
          <input 
            type='text' 
            placeholder='Enter Name'
            name='name'
            value={this.state.name}
            onChange={this.handleChange}
          />
          <input 
            type='text' 
            placeholder='Enter Email'
            name='email'
            value={this.state.email}
            onChange={this.handleChange}
          />
          <input 
            type='text'  
            placeholder='Enter Password'
            name='password'
            value={this.state.password}
            onChange={this.handleChange}
          />
          <Link to='/login'>
            <button 
              className='login-btn' 
              type='button'
              onClick={this.createAccount} 
            >Create Account</button>
          </Link>
        </form>
      </section>
    )
  }
}

export default CreateAccount;