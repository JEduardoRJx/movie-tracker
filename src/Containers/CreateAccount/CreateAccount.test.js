import React from 'react';
import { shallow } from 'enzyme';
import CreateAccount from './CreateAccount';
import { addUser } from '../../utils/apiCalls';
jest.mock('../../utils/apiCalls');

describe('CreateAccount', () => {
  let wrapper;
  const mockNameEvent = {
    target: {
      name: 'name',
      value: 'Amy'
    }
  };
  const mockEmailEvent = {
    target: {
      name: 'email',
      value: 'amy@email.com'
    }
  };
  const mockPasswordEvent = {
    target: {
      name: 'password',
      value: 'password'
    }
  };

  beforeEach(() => {
    wrapper = shallow(<CreateAccount />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when change happens', () => {
    wrapper.instance().handleChange(mockNameEvent);
    expect(wrapper.state('name')).toEqual('Amy');
    wrapper.instance().handleChange(mockEmailEvent);
    expect(wrapper.state('email')).toEqual('amy@email.com');
    wrapper.instance().handleChange(mockPasswordEvent);
    expect(wrapper.state('password')).toEqual('password');
  });

  it('should reset state when clearInputs is called', () => {
    const mockState = {
      name: 'Amy',
      email: 'amy@email.com',
      password: 'password',
      error: '',
      success: false
    };
    const expected = {
      name: '',
      email: '',
      password: '',
      error: '',
      success: false
    }
    wrapper.instance().setState(mockState);
    expect(wrapper.state()).toEqual(mockState );
    wrapper.instance().clearInputs();
    expect(wrapper.state()).toEqual(expected);
  });

  it('should call createAccount when button is clicked', () => {
    wrapper.instance().createAccount = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().createAccount).toHaveBeenCalled();
  });
});