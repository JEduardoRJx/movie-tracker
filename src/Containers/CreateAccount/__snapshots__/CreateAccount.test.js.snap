// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount should match the snapshot 1`] = `
<section
  className="form-section"
>
  <form
    className="create-account-form"
  >
    <h2
      className="create-account-h2"
    >
      Create Account: 
    </h2>
    <input
      name="name"
      onChange={[Function]}
      placeholder="Enter Name"
      type="text"
      value=""
    />
    <input
      name="email"
      onChange={[Function]}
      placeholder="Enter Email"
      type="text"
      value=""
    />
    <input
      name="password"
      onChange={[Function]}
      placeholder="Enter Password"
      type="text"
      value=""
    />
    <button
      className="login-btn"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Create Account
    </button>
  </form>
</section>
`;
